ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,450 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
 
 static
 NTSTATUS
-OpenRegistryHandlesFromSymbolicLink(IN PUNICODE_STRING SymbolicLinkName,
-                                    IN ACCESS_MASK DesiredAccess,
-                                    IN OPTIONAL PHANDLE GuidKey,
-                                    IN OPTIONAL PHANDLE DeviceKey,
-                                    IN OPTIONAL PHANDLE InstanceKey)
+IopBuildSymbolicLink(IN PUNICODE_STRING PrefixString,
+                     IN PUNICODE_STRING DeviceString,
+                     IN PUNICODE_STRING GuidString,
+                     IN OPTIONAL PUNICODE_STRING ReferenceString,
+                     OUT PUNICODE_STRING SymbolicLinkName)
 {
-    OBJECT_ATTRIBUTES ObjectAttributes;
-    UNICODE_STRING BaseKeyU;
-    UNICODE_STRING GuidString, SubKeyName, ReferenceString;
-    PWCHAR StartPosition, EndPosition;
-    HANDLE ClassesKey;
-    PHANDLE GuidKeyRealP, DeviceKeyRealP, InstanceKeyRealP;
-    HANDLE GuidKeyReal, DeviceKeyReal, InstanceKeyReal;
+    UNICODE_STRING SymbolicLink;
     NTSTATUS Status;
 
-    SubKeyName.Buffer = NULL;
+    /* Build up new symolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = PrefixString->Length + DeviceString->Length + sizeof(L"#") + GuidString->Length + ReferenceString->Length;
+    SymbolicLink.Buffer = ExAllocatePoolWithTag(NonPagedPool, SymbolicLink.MaximumLength, TAG_IO);
+    if (!SymbolicLink.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        ExFreePoolWithTag(SymbolicLink.Buffer, TAG_IO);
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
 
-    if (GuidKey != NULL)
-        GuidKeyRealP = GuidKey;
-    else
-        GuidKeyRealP = &GuidKeyReal;
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, PrefixString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
 
-    if (DeviceKey != NULL)
-        DeviceKeyRealP = DeviceKey;
-    else
-        DeviceKeyRealP = &DeviceKeyReal;
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, DeviceString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
 
-    if (InstanceKey != NULL)
-        InstanceKeyRealP = InstanceKey;
+    Status = RtlAppendUnicodeToString(&SymbolicLink, L"#");
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    if (ReferenceString->Buffer != NULL)
```

Comment:
Please check `Length` to look for an empty string, not `Buffer`

Reply:
Now I'm checking whether the string pointer is `NULL` instead, since this parameter actually should be optional.